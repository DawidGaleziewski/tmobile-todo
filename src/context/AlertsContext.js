import React, { createContext, useState } from "react";

// Hooks
import useAlerts from "../hooks/useAlerts";

// Utils
import { v4 as uuidv4 } from "uuid";

const AlertsContext = createContext(null);

const AlertsProvider = ({ children }) => {
  const [alerts, setAlerts] = useState([]);
  const addAlert = (alertType, alertText) => {
    const id = uuidv4();
    const updatedAlerts = [
      ...alerts,
      {
        text: alertText,
        type: alertType,
        id: id,
      },
    ];
    setAlerts(updatedAlerts);
  };
  const removeAlert = (id) => {
    setAlerts([...alerts].filter((alert) => alert.id !== id));
  };

  useAlerts(alerts, removeAlert);

  return (
    <AlertsContext.Provider
      value={{
        alerts,
        addAlert,
        removeAlert,
        setAlerts,
      }}
    >
      {children}
    </AlertsContext.Provider>
  );
};

export { AlertsContext, AlertsProvider };
