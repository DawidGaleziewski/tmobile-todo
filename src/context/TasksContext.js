import React, { createContext, useState } from "react";

// Hooks
import useAxiosGetCall from "../hooks/useAxiosGetCall";

// Authorization
import { tasksAPIUrl } from "../auth/tasksAPISettings";

const TasksContext = createContext([]);

const TasksProvider = ({ children, setIsloading }) => {
  const [tasksState, setTasksState] = useState([]);

  const updateOneTaskState = (itemID, newItem) => {
    const findItemIndex = (array, itemID) =>
      array.findIndex((item) => item.id === itemID);

    const getUpdatedArray = (array, itemID, newItem) => {
      const arrayCopy = JSON.parse(JSON.stringify(array));
      arrayCopy.splice(findItemIndex(arrayCopy, itemID), 1, newItem);
      return arrayCopy;
    };

    setTasksState(getUpdatedArray(tasksState, itemID, newItem));
  };

  const removeTask = (taskID) => {
    console.log(taskID);
    setTasksState(
      [...tasksState].filter((task) => task.id !== parseInt(taskID))
    );
    console.log(tasksState);
  };

  useAxiosGetCall(
    tasksAPIUrl,
    setIsloading,
    (data) => {
      setTasksState(data);
    },
    []
  );

  return (
    <TasksContext.Provider
      value={{ tasksState, setTasksState, updateOneTaskState, removeTask }}
    >
      {children}
    </TasksContext.Provider>
  );
};

export { TasksContext, TasksProvider };
