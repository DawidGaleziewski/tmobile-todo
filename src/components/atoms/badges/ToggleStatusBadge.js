import React, { useState, useContext } from "react";

// Components
import LoadingCircle from "../loadings/LoadingCircle";

// Libraries
import { css, jsx } from "@emotion/core";

// Context
import { AlertsContext } from "../../../context/AlertsContext";
import { TasksContext } from "../../../context/TasksContext";

// Utils
import editTaskCall from "../../../utils/httpAxiosCalls/editTaskCall";

// Authorization
import { tasksAPIUrl } from "../../../auth/tasksAPISettings";

// Styles
import { badgeStyle } from "./style/ToggleStatusBadgeStyle";

/** @jsx jsx */

const ToggleStatusBadge = ({ task }) => {
  const [isLoading, setIsloading] = useState(false);
  const { updateOneTaskState } = useContext(TasksContext);
  const { addAlert } = useContext(AlertsContext);
  const { id, done } = task;

  const singleTaskApiUrl = `${tasksAPIUrl}/${id}`;

  const onClickHandler = () => {
    const newTaskData = {
      done: !done,
    };
    editTaskCall(
      singleTaskApiUrl,
      newTaskData,
      setIsloading,
      addAlert,
      (updatedTaskData) => {
        updateOneTaskState(updatedTaskData.id, updatedTaskData);
      }
    );
  };

  return isLoading ? (
    <LoadingCircle />
  ) : (
    <span
      data-testid="toggle-status-badge"
      onClick={onClickHandler}
      css={badgeStyle(done)}
    >
      {done ? "done" : "in progress"}
    </span>
  );
};
export default ToggleStatusBadge;
