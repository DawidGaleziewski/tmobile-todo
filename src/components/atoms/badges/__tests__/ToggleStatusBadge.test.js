// Libraries
import React, { useContext } from "react";
import ReactDOM from "react-dom";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";

// Components
import ToggleStatusBadge from "../ToggleStatusBadge";

// Context
import { AlertsProvider } from "../../../../context/AlertsContext";

afterEach(cleanup);

it("renders ToggleStatusBadge without crashing", () => {
  const divContainer = document.createElement("div");
  const taskMockData = {
    id: 20,
    done: true,
    title: "test",
    description: "test",
  };
  ReactDOM.render(
    <AlertsProvider>
      <ToggleStatusBadge task={taskMockData} />
    </AlertsProvider>,
    divContainer
  );
  ReactDOM.unmountComponentAtNode(divContainer);
});

it("renders ToggleStatusBadge with correct text", () => {
  const taskMockData = {
    id: 20,
    done: true,
    title: "test",
    description: "test",
  };
  const { getByTestId } = render(
    <AlertsProvider>
      <ToggleStatusBadge task={taskMockData} />
    </AlertsProvider>
  );
  expect(getByTestId("toggle-status-badge")).toHaveTextContent("done");
});

it("matches snapshot 1", () => {
  const taskMockData = {
    id: 20,
    done: false,
    title: "test 2",
    description: "test2",
  };
  const DOMTree = renderer
    .create(
      <AlertsProvider>
        <ToggleStatusBadge task={taskMockData} />
      </AlertsProvider>
    )
    .toJSON();

  expect(DOMTree).toMatchSnapshot();
});
