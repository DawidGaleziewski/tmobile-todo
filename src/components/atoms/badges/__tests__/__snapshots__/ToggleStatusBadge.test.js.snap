// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1 1`] = `
<span
  className="css-mozazn-ToggleStatusBadge"
  data-testid="toggle-status-badge"
  onClick={[Function]}
>
  in progress
</span>
`;
